### ISSUE TITLE
[bug] conan create --json content references are not suitable for upload in user _ channel _ case

### ISSUE THREAD

**Issue by michaelmaguire** [conan-io/conan#6862]
### Environment Details (include every applicable attribute)
  * Operating System+version Windows Server 2012 R2
  * Compiler+version: vs2015
  * Conan version: Conan version 1.19.2
  * Python version: N/A

### Steps to reproduce (Include if Applicable)

At my org, we can't use pre-built binaries from outside, but we can use  conan-center-index recipes.  I'm trying to build https://github.com/conan-io/conan-center-index/blob/master/recipes/7zip/19.00/conanfile.py#L37 but it references: 

```
   self.build_requires("make/4.2.1")
```

I already have make/4.2.1@myorg/stable in my internal Artifactory, but Conan won't pick this up.  

I asked in cpplang.slack

> What's the right way to get default USER/CHANNEL working for this recipe?

and it was suggested:

> Hipony:unicorn_face:  17:44
> You need to override it. I'd consider it a mistake that you're using your own user/channel for recipes that are intended to be an automatic replacement for all existing recipes. If you'd vendor it into your own Artifactory as just make/4.2.1 it would pick it up.

I tried using our CI to upload make/4.2.1 by specifying user `_` channel `_` and it built:

```
16:58:26 conan_package_user[_]
[Pipeline] echo
conan_package_channel[_]
16:59:04 make/4.2.1: Package '456f15897172eef340fcbac8a70811f2beb26a93' built
```

but then later when our CI tries to upload to our internal Artifactory, I get the following error:

```
[all] $ cmd.exe /C "conan upload make/4.2.1#2516c4ecfaf0129339951b0be88ed356 --all -r b9015ace-7d9e-46a3-9764-75b75e176f26 --confirm --no-overwrite recipe && exit %%ERRORLEVEL%%"
DEBUG :conan_api.py   [163]: INIT: Using config 'E:\nightlybuild\workspace\Conan_pipeline\uh\.conan\conan.conf' [2020-04-15 10:56:44,322]
DEBUG :tracer.py      [155]: CONAN_API: upload(pattern=make/4.2.1#2516c4ecfaf0129339951b0be88ed356,package=None,query=None,remote_name=b9015ace-7d9e-46a3-9764-75b75e176f26,all_packages=True,policy=no-overwrite-recipe,confirm=True,retry=None,retry_wait=None,integrity_check=False) [2020-04-15 10:56:44,325]
ERROR: No packages found matching pattern 'make/4.2.1#2516c4ecfaf0129339951b0be88ed356'
```

It was suggested that the error is a missing `@`:

> Hipony:unicorn_face:  17:52 Log is a little bit confusing
> 17:53 aha
> 17:53 use make/4.2.1@#2516c4ecfaf0129339951b0be88ed356
> 17:53 note @ in the reference

However, the: 'make/4.2.1#2516c4ecfaf0129339951b0be88ed356' is the conan reference I parsed out of the json from the conan create --json command

i.e. I never created the string `make/4.2.1@#2516c4ecfaf0129339951b0be88ed356`. 

Instead, I called `conan create` with the `--json` option:

```
conanClient.run(buildInfo: buildInfo, command: "create . ${conan_package_version}@${conan_package_user}/${conan_package_channel} --profile ${it} --json conanCreateOutput_${it}.json ${conan_additional_create_parameters}".toString())
```

Then, to ensure we only attempt to upload to our internal Artifactory only packages we just built in this build job, we then parse all packages with 'built' set:

```
                        def builtPackages = created_refs(conanCreateOutputJson)
```

using:

```
// Returns a list with full package id of all Conan
// packages actually built (not downloaded), by parsing
// the --json file return from the conan create command.
def static created_refs(a_json){
    def refs = []
    a_json["installed"].each { installed ->
    def recipe = installed['recipe']
    def packages = installed['packages']
    def atLeastOnePackageOfRecipeWasBuilt = false
    if( packages != null ) {
        packages.each { thePackage ->
            def built = thePackage['built']
            if( built != null && built ) {
                atLeastOnePackageOfRecipeWasBuilt = true
            }
        }
    }
    // Only add one copy of the package id, no
    // matter how many binary packages for it were built.
    if( atLeastOnePackageOfRecipeWasBuilt ) {
        def id = recipe['id']
        refs << id
    }
    }
    return refs
} 
```

and then finally we perform a `conan upload` for only those packages that were built:

```
                            builtPackages.each {
                            
                                env.MY_FAILURE_STAGE = STAGE_NAME + "__loop_upload_packages_built_${it}"
                            
                                def command = "upload ${it} --all -r ${artifactoryUploadRemote} ${conan_additional_upload_parameters}"
                                conanClient.run buildInfo: buildInfo, command: command.toString()
                            }
```

So I'm basically calling  `conan upload` on `${it}` where contents of ${it} came from the json returned from conan create.

I would expect that `conan create --json` content should contain references in a format appropriate for passing to the `conan upload` command, even in cases where user and channel are `_`.

**Comment by michaelmaguire** [comment]
I note that if I perform a slight hack in my Jenkinsfile in order to follow Hipony's advice :

```
				// Ensure we have a valid package reference for the upload.
				// Work around https://github.com/conan-io/conan/issues/6862
				if( builtPackage.indexOf('@') == -1 ) {
                            		echo "builtPackage does not contain @"
					if( builtPackage.indexOf('#') == -1 ) {
                            			echo "builtPackage does not contain #, appending @"
						builtPackage = builtPackage + "@"
					} else {
                            			echo "builtPackage does contain #, inserting @ immediately before it"
						builtPackage = builtPackage.replaceFirst("#", "@#")
					}
				}
```


then it works:

```
builtPackage[make/4.2.1#2516c4ecfaf0129339951b0be88ed356]
[Pipeline] echo
builtPackage does not contain @
[Pipeline] echo (hide)
builtPackage does contain #, inserting @ immediately before it
[Pipeline] runConanCommand
[all] $ cmd.exe /C "conan upload make/4.2.1@#2516c4ecfaf0129339951b0be88ed356 --all -r fcf0d8bc-57eb-4aec-8eed-31eea706dade --confirm --no-overwrite recipe && exit %%ERRORLEVEL%%"
DEBUG :conan_api.py   [163]: INIT: Using config 'E:\nightlybuild\workspace\ALPHA_test_Conan_pipeline\uh\.conan\conan.conf' [2020-04-15 14:07:04,981]
DEBUG :tracer.py      [155]: CONAN_API: upload(pattern=make/4.2.1@#2516c4ecfaf0129339951b0be88ed356,package=None,query=None,remote_name=fcf0d8bc-57eb-4aec-8eed-31eea706dade,all_packages=True,policy=no-overwrite-recipe,confirm=True,retry=None,retry_wait=None,integrity_check=False) [2020-04-15 14:07:04,984]
DEBUG :rest_client_common.py[160]: REST: ping: https://mycorpartprod/artifactory/api/conan/bb-conan-nx-stable-local/v1/ping [2020-04-15 14:07:05,017]
Uploading to remote 'fcf0d8bc-57eb-4aec-8eed-31eea706dade':
Uploading make/4.2.1 to remote 'fcf0d8bc-57eb-4aec-8eed-31eea706dade'
Compressing recipe...
[==================================================]


Compressing recipe sources...
[==================================================]
```

**Comment by jgsogo** [comment]
Time ago we implemented a feature to allow packages without user/channel, their main intention is to use it for packages in ConanCenter.... but there is a problem in the command line we cannot resolve before breaking in Conan 2.0. There are a lot of commands that accept a reference or a path, and the `@` at the end is the only way to know the user refers to a reference (without it, Conan will keep thinking it is a path):

```
conan install thing/thing  <--- this will look for a conanfile in `thing/thing/conanfile.py`
conan install thing/thing@  <--- this is a reference (without user/channel)
conan install thing/thing@user/channel  <--- this is a reference
```

Regarding the JSON output, we need to look into this issue and think if we can change the JSON output so it can be used directly in other commands. Today I found the same issue trying to search a specific revisions, it needs to be `conan search name/version@#revision`. Maybe we can get rid of that `@` without breaking anything (not sure if it something intended or not)
